### DISCUSSION TITLE
how to specify columns with loadCSV?

### DISCUSSION THREAD

**Discussion by fundef1** [uwdata/mosaic#181] [Q&A] [answered]
Hi, I'm trying to load a CSV and manually specify the columns.
it's a bit of a degenerate case as I only have a single column, but I can't figure out the correct format:

```
const csvopts = {
	auto_detect: false, 
	all_varchar: true,
	columns: "{'line': 'VARCHAR'}", 
	header: false
};

vg.coordinator().exec(vg.loadCSV("lines", `${window.location}data/lines.csv`,csvopts));
```
results in 
```
Uncaught (in promise) Error: Parser Error: syntax error at or near "line"
LINE 1: ..._size=-1, all_varchar=true, columns='{'line': 'VARCHAR'}', header=false,...
```
the quoting of columns looks off and  I've tried different variants of quoting the columns option but just can't get it to work.

any hints? thanks!

**Comment by jheer** [comment]
The `loadCSV` method supports a `select` option, as described in the docs here: https://uwdata.github.io/mosaic/api/sql/data-loading.html#loadcsv

For example: `vg.loadCSV(table, filename, { select: ['ColumnA', 'ColumnB'] })`

**Reply by fundef1** [reply]
Right, but that's not what I'm after. I do not want to select a subset of columns. 
DuckDB requires that you specify the columns with auto_detect false: 
```
Additional CSV-specific options, see the [DuckDB CSV documentation](https://duckdb.org/docs/data/csv/overview).
```
and in the table with csv options from duckdb:
<html>
<body>


columns | A struct that specifies the column names and column types contained within the CSV file (e.g., {'col1': 'INTEGER', 'col2': 'VARCHAR'}). | struct | (empty)
-- | -- | -- | --



</body>
</html>

it's this struct that I need to pass to duckdb.

**Reply by jheer** [reply, marked as answer]
Apologies, I triaged too quickly and misunderstood. This is a shortcoming, as input string values will always be quoted. We need to expand the option processor to support objects (-> structs) and arrays as well. I will create an issue for this.

**Reply by jheer** [reply]
Fixed by https://github.com/uwdata/mosaic/pull/183
